src/main/scala/rise/core/substitute.scala

```diff
@@ -28,7 +28,12 @@ object substitute {
         if (`for` == e) {
           Stop(expr)
         } else {
-          Continue(e, this)
+          e match {
+            case Lambda(x, _)  if x == `for` =>
+              // identified shadowing => do not recurse the substitution
+              Stop(e)
+            case _ => Continue(e, this)
+          }
```

Comment:
I do not think that this change is enough to support arbitrary substitution without unique names? See capture-avoiding substitution: https://www.cs.cornell.edu/courses/cs3110/2019sp/textbook/interp/subst_lambda.html

Reply:
I have updated the implementation.